File: py/p10.py
import math
from typing import List

def sieve_up_to(limit: int) -> List[bool]:
    if limit <= 3:
        return [False]

    max_odd = limit - 1 - (limit % 2)
    sieve = [False] + [True] * (max_odd // 2)
    cross = math.floor(math.sqrt(max_odd)) // 2 + 1

    for i in range(1, cross):
        if sieve[i]:
            for j in range(2*i*(i+1), len(sieve), 2*i+1):
                sieve[j] = False

    return sieve

def sum_of_primes(sieve: List[bool]) -> int:
    sum = 2
    for i in range(len(sieve)):
        if sieve[i]:
            sum += 2 * i + 1
    return sum

limit = 2000000
print(sum_of_primes(sieve_up_to(limit)))

File: rs/p10.rs
fn main() {
    const LIMIT: usize = 2000000;
    let sieve = sieve_up_to(LIMIT);
    let sum = sum_of_primes(&sieve);
    println!("{}", sum);
}

fn sieve_up_to(limit: usize) -> Vec<bool> {
    if limit <= 3 {
        vec![false]
    } else {
        let max_odd = limit - 1 - (limit % 2);
        let length = max_odd / 2 + 1;
        let cross = ((max_odd as f64).sqrt().floor() as usize) / 2 + 1;

        let mut sieve = vec![true; length];
        sieve[0] = false;
        for i in 1..cross {
            if sieve[i] {
                for j in ((2 * i * (i + 1))..length).step_by(2*i+1) {
                    sieve[j] = false
                }
            }
        }
        sieve
    }
}

fn sum_of_primes(sieve: &Vec<bool>) -> u128 {
    sieve.into_iter().enumerate().fold(2u128, |s, (i, is_prime)| if *is_prime {
        s + 2 * (i as u128) + 1
    } else {
        s
    })
}
